components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/definitions/OASParserUtil.java

```diff
@@ -655,6 +668,74 @@ private static String getComponentCategory(String ref) {
         return "";
     }
 
+    /**
+     * Extract the archive file and validates the openAPI definition
+     *
+     * @param inputStream   file as input stream
+     * @param returnContent whether to return the content of the definition in the response DTO
+     * @return APIDefinitionValidationResponse
+     * @throws APIManagementException if error occurred while parsing definition
+     */
+    public static APIDefinitionValidationResponse extractAndValidateOpenAPIArchive(InputStream inputStream,
+            boolean returnContent) throws APIManagementException {
+        String path = System.getProperty(APIConstants.JAVA_IO_TMPDIR) + File.separator +
+                APIConstants.OPENAPI_ARCHIVES_TEMP_FOLDER + File.separator + UUID.randomUUID().toString();
+        String archivePath = path + File.separator + APIConstants.OPENAPI_ARCHIVE_ZIP_FILE;
+        String extractedLocation = APIFileUtil
+                .extractUploadedArchive(inputStream, APIConstants.OPENAPI_EXTRACTED_DIRECTORY, archivePath, path);
+        File[] listOfFiles = new File(extractedLocation).listFiles();
+        File archiveDirectory = null;
+        if (listOfFiles != null) {
+            for (File file: listOfFiles) {
+                if (file.isDirectory()) {
+                    archiveDirectory = file.getAbsoluteFile();
+                }
+            }
+        }
+        if (archiveDirectory == null) {
+            throw new APIManagementException("Could not find an archive in the given ZIP file.");
+        }
+        File masterSwagger;
+        if ((new File(archiveDirectory + "/" + APIConstants.OPENAPI_MASTER_JSON)).exists()) {
+            masterSwagger = new File(archiveDirectory + "/" + APIConstants.OPENAPI_MASTER_JSON);
+        } else if ((new File(archiveDirectory + "/" + APIConstants.OPENAPI_MASTER_YAML)).exists()) {
+            masterSwagger = new File(archiveDirectory + "/" + APIConstants.OPENAPI_MASTER_YAML);
+        } else {
+            throw new APIManagementException("Could not find a master swagger file with the name of swagger.json "
+                    + "/swagger.yaml");
+        }
+        String content;
+        try {
+            InputStream masterInputStream = new FileInputStream(masterSwagger);
+            content = IOUtils.toString(masterInputStream, APIConstants.CHARSET);
+        } catch (IOException e) {
+            throw new APIManagementException("Error reading master swagger file");
+        }
+        String openAPIContent = "";
+        SwaggerVersion version;
+        version = getSwaggerVersion(content);
+        if (SwaggerVersion.OPEN_API.equals(version)) {
+            String filePath = masterSwagger.getAbsolutePath();
+            OpenAPIV3Parser openAPIV3Parser = new OpenAPIV3Parser();
+            ParseOptions options = new ParseOptions();
```

Comment:
can we define another method named "validateAPIDefinitionByPath" on respective OAS parsers? Then you can call it from here.
if so you can simply return openAPIV3Parser.validateAPIDefinitionByPath(path, returnJsonContent) here.

in respective parsers, you have to implement validateAPIDefinitionByPath function with following specific logic to parser the file path. In there you can set the correct OAS model(OpenAPI or Swagger) to the APIDefinitionValidationResponse.

With that simply you can reduce the unnecessary content parsing in multiple places.

Reply:
What do you mean by unnecessary content parsing ?

Here are creating a parser, resolving and getting the specific parser object and converting the parser object to string. These three folds are specific for each parsers and even if we have a separate method logic the same should be done in method logic too.

Reply:
Since are parsing file multiple times it is better to avoid unnecessary parsing. Also, it is better to separate OAS3 and OAS2 to respective classes. If we support a naming convention to identify master swagger, this can be done.
Can you please check the attached diff? [OAS_Path_validation.txt](https://github.com/wso2/carbon-apimgt/files/4640269/OAS_Path_validation.txt)